### ISSUE TITLE
Consolidate methods to configure attribute parsing in macro processors

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#2122]
Inspired by #2121, consolidate the methods on the macro processor extensions for configuring attribute parsing. Currently, there are three methods:

* content_model (aliased as parse_content_as)
* positional_attributes (aliased as name_positional_attributes, name_attributes, and pos_attrs)
* default_attrs (aliased as seed_attributes_with)

These methods feel clumsy. They can be reduced to a single method named `resolves_attribute` that can handle the following settings:

1. enable attribute parsing
2. map / bind positional attributes to attribute names
3. assign default attribute values

Here's an example that shows how this method can be used to disable attribute parsing:

```ruby
resolves_attributes false
```

(In this case, the content between the square brackets gets stored in the attribute named text).

Here's an example that shows how to enable attribute parsing:

```ruby
resolves_attributes
```

Here are examples that shows how to map positional attributes to attribute names:

```ruby
resolves_attributes 'size'
```

```ruby
resolves_attributes 'alt', 'width', 'height'
```

Here's an example that shows how to specify default attributes:

```ruby
resolves_attributes '@size' => 5, 'max-word-size' => 10
```

(The leading @, which gets dropped, marks an attribute name as positional).

**Comment by mojavelinux** [comment]
Here are some examples that shows this new method in action:

```ruby
# ex. tel:1-800-555-1212[Call Now]
Asciidoctor::Extensions.register do
  inline_macro do
    named :tel
    resolves_attributes false
    process do |parent, target, attrs|
      puts %(tel: #{target}; label: #{attrs['text']})
      nil
    end
  end
end
```

```ruby
# ex. lorem::sentences[3,max-word-size=10]
Asciidoctor::Extensions.register do
  block_macro do
    named :lorem
    resolves_attributes '@size': 1, 'max-word-size': 5
    process do |parent, target, attrs|
      puts %(generate #{attrs['size']} #{target} with a max word size of #{attrs['max-word-size']})
      nil
    end
  end
end
```

This feels really natural to me.

**Comment by mojavelinux** [comment]
@Mogztter wdyt about this proposal?

**Comment by mojavelinux** [comment]
The other idea I had to express that a default attribute key is also a positional attribute is to surround the name in square brackets:

```ruby
resolves_attributes '[size]': 1, 'max-word-size': 5
```

So it's clear, `max-word-size` is not a positional attribute. It just has a default value.

**Comment by mojavelinux** [comment]
Somewhat related to this issue, I'm thinking of using the name `_text` to store the unparsed attributes (which we'd keep even when parsing them). If attribute parsing is disabled, then the attributes hash only contains one key, `_text`.

(Another name I had in mind is `attrlist`, inspired by the keyword that AsciiDoc Python uses).

**Comment by mojavelinux** [comment]
And yet another name I had in mind is `macrotext`, though I like that one a bit less.

**Comment by Mogztter** [comment]
That's a HUGE improvement!
I didn't really understand how to use `content_model`, `positional_attributes` and `default_attrs`.

I think it can be confusing to "reuse" the `@` symbol for positional attributes. I think it's better to surround the name in square brackets.

I do not like `macrotext`. `_text` is OK. What about `_raw` ? unparsed content = raw content.

I'm looking forward to use this new method, well done :+1:

**Comment by mojavelinux** [comment]
Indeed, I can really sense things coming together.

I agree the `@` symbol looks a bit vague. I have another proposal that's more explicit. The position is specified using a 1-based number followed by a colon.

```ruby
resolves_attributes '1:size': 5, 'max-word-size': 5
```

This syntax isn't used that often, and typically you only have a few positional attributes, so it's easy to track. We could even reintroduce the `@` again to represent the next number in the sequence.

```ruby
resolves_attributes '@:size': 5, 'max-word-size': 5
```

But I'll start with the explicit position number and we'll evolve from there.

**Comment by mojavelinux** [comment]
> I do not like macrotext. _text is OK. What about _raw ? unparsed content = raw content.

As much as possible I try to avoid the word "raw" in Asciidoctor because it's has a very vague meaning.

This really comes down to how we document the macro. Let's assume we describe is as follows:

```
<name>:<target>[<text|attrlist>]
```

In that case, "text" makes sense. I shy away from "attrlist" to represent the unparsed text because the value is either attributes or it's not. "attrlist" brings the question to mind "does that mean the unparsed text or the actual attributes?" So when we say "attrlist", we really mean "parsed attributes".

With that logic, perhaps we should document it as:

```
<name>:<target>[<text|attributes>]
```

(In this case, the keyword between the square brackets matches the content model).

I was going to suggest "content", but attributes are type of content and that's too vague. In XML, text is assumed to be character data and markup, and that's really what we are shooting for.

If we tried to follow the terminology in the XML spec, the alternative to text would be "data". But that doesn't feel quite right to me.

**Comment by mojavelinux** [comment]
So the open question is are "text" and "attributes" the appropriate keywords to represent unparsed and parsed attributes?

**Comment by Mogztter** [comment]
`attributes` (or `attrs`) is also the name of the variable:

```ruby
process do |parent, target, attrs|
  # attrs._text = textual representation of attributes (unparsed)
  # attrs['max-word-size'] = max-word-size attribute
end
```

Asking for the `text` (or the `raw`) representation of the attributes makes sense for me.

> However, I shy away from "attrlist" to represent the unparsed text because the value is either attributes or it's not. "attrlist" brings the question to mind "does that mean the unparsed text or the actual attributes?" So when we say "attrlist", we really mean "parsed attributes".

I completely agree.

> So the open question is are "text" and "attributes" the appropriate keywords to represent unparsed and parsed attributes?

Yes.
"text" should be prefixed by an underscore to avoid conflict with an attribute named "text" ?

**Comment by mojavelinux** [comment]
Thanks for the input!

I remembered that we don't need to worry about a name conflict since the "text" attribute is only set when attributes are not parsed. So when it's set, it's the only attribute.

There are also a lot of extensions out there that rely on the "text" attribute, so we'd have to keep it for backwards compatibility. We could, however, also set the "_text" attribute and encourage people to change. But at this point, I don't see much motivation for doing so.

There is still an open question about whether we should always make the text between brackets available. If so, I like your idea of making it a field on the attributes object. That kind of hides it away a bit. But I think we should deal with that separately.

**Comment by mojavelinux** [comment]
I'm a bit torn whether to name this method "resolves_attributes" or "resolve_attributes". Here's my thinking about this. For DSL methods that bind the processor to the syntax, we use adjectives.

* named (e.g., `named :label`)
* with_form (e.g., `with_form :short`)
* matching (e.g., `matching /@(\w+)/`)

For DSL methods that instruct the processor how to behave, we use commands:

* resolve_attributes
* parse_content_as

**Comment by mojavelinux** [comment]
I was thinking more about mapping positional attributes using resolve_attributes and I'd like to propose a slight change. Consider yourself coming into a project and seeing the following:

```ruby
resolve_attributes 'alt', 'width', 'height'
```

That might lead you to believe that the only attributes being resolve are "alt", "width", and "height". It's not clear from that syntax that those are the positional attributes. Therefore, I think we should use the positional mapping syntax we introduced for default attributes in this case as well.

```ruby
resolve_attributes '1:alt', '2:width', '3:height'
```

I was playing around with the syntax and while I like the colon, I think an equal sign is even more clear:

```ruby
resolve_attributes '1=alt', '2=width', '3=height'
```

(We could allow either form).

What's nice about this syntax is that it plays right into adding a default value:

```ruby
resolve_attributes '1=alt' => 'figure', '2=width' => 800, '3=height' => 600
```

However, the equal sign might cause some confusion there, so back to the colon:

```ruby
resolve_attributes '1:alt' => 'figure', '2:width' => 800, '3:height' => 600
```

In that case, we could even allow the default to be set when using the array form:

```ruby
resolve_attributes '1:alt=figure', '2:width=800', '3:height=600'
```

That seems to capture the meaning in a very concise and clear way. It's easy to drop the default value or to add a default to a non-positional attribute:

```ruby
resolve_attributes '1:alt', '2:width=800', '3:height=600', 'format=png'
```

**Comment by mojavelinux** [comment]
tl;dr here's the final proposal using the array shorthand:

```ruby
resolve_attributes '1:alt', '2:width=800', '3:height=600', 'format=png'
```

we could also accept that as a hash:

```ruby
resolve_attributes '1:alt'=>nil, '2:width'=>800, '3:height'=>600, 'format'=>'png'
```

That encapsulates:

* enabling attribute parsing (setting the content_model option to `:attributes`)
* assigning (key) names to positional attributes (setting the pos_attrs option)
* setting default attributes (setting the default_attrs option)

**Comment by mojavelinux** [comment]
The groundwork is in place. I'm ready to start hacking on the PR for this one.
